core/3d/CCMeshVertexIndexData.cpp

```diff
@@ -132,14 +132,21 @@ MeshVertexData* MeshVertexData::create(const MeshData& meshdata)
     bool needCalcAABB = (meshdata.subMeshAABB.size() != meshdata.subMeshIndices.size());
     for (size_t i = 0, size = meshdata.subMeshIndices.size(); i < size; ++i)
     {
-        auto& index      = meshdata.subMeshIndices[i];
+        auto& index = meshdata.subMeshIndices[i];
+        auto indexSize = format == CustomCommand::IndexFormat::U_SHORT ? sizeof(uint16_t) : sizeof(uint32_t);
         auto indexBuffer = backend::Device::getInstance()->newBuffer(
-            index.size() * sizeof(index[0]), backend::BufferType::INDEX, backend::BufferUsage::STATIC);
+            index.size() * indexSize, backend::BufferType::INDEX, backend::BufferUsage::STATIC);
         indexBuffer->autorelease();
 #if CC_ENABLE_CACHE_TEXTURE_DATA
         indexBuffer->usingDefaultStoredData(false);
 #endif
-        indexBuffer->updateData((void*)index.data(), index.size() * sizeof(index[0]));
+        if (format == CustomCommand::IndexFormat::U_SHORT)
+        {
+            std::vector<unsigned short> shortIndexFormat(index.begin(), index.end());
```

Comment:
is possible avoid copy the indices?

Reply:
You have to convert the data in the end no matter what

Reply:
You could avoid it but if the user is unaware it would generate artifacts, Do you think it should be avoided?

Reply:
maybe, we can wrap IndexArray as a class extend byte_buffer and write two constructors with argument initialize_list uint and ushort

Reply:
Good Idea

Reply:
How about this?

```
/** Index array reposible for mesh index formatting and handeling.
 * choose either uint16_t (U_SHORT) or uint32_t (U_INT) for the template name
 */
struct GenericIndexArray
{
    std::vector<unsigned short> _ushortIndices;
    std::vector<unsigned int> _uintIndices;

    CustomCommand::IndexFormat _format;

public:
    template <typename T>
    std::vector<T>& getArray()
    {
        _format = sizeof(T) == 2 ? CustomCommand::IndexFormat::U_SHORT : CustomCommand::IndexFormat::U_INT;

        if (_format == CustomCommand::IndexFormat::U_SHORT)
            return _ushortIndices;
        if (_format == CustomCommand::IndexFormat::U_INT)
            return _uintIndices;
    }
};
```

Reply:
For example:
```cpp
class IndexArray : public yasio::byte_buffer {
public:
  IndexArray(const std::initializer_list<unsigned int>& rhs) 
  { 
      this->assign((const uint8_t*)rhs.begin(), (const uint8_t*)rhs.end());
  }

  IndexArray(const std::initializer_list<unsigned short>& rhs) 
  { 
      this->assign((const uint8_t*)rhs.begin(), (const uint8_t*)rhs.end());
  }
};
```

Reply:
isn't is still copying memory here? also how can we deal with the data in the byte_buffer in opengl and actually tell the gpu how to read the data?

Reply:
```cpp
Because user decide the indexArray fomat and fill uint or ushort array to the IndexArray, you can updataData with index array directly with indexForamt enum
```

Reply:
Honestly I don't think it has a solution but to copy, since you're copying anyway with byte_buffer class from unsigned char to other types, So I think the conversion function should stay `std::vector<unsigned short> shortIndexFormat(index.begin(), index.end());` and also It's only copied when being sent to the gpu only

Reply:
or you can copy once without a temp `std::vector<unsigned short>`

Reply:
> For example:
> 
> ```c++
> class IndexArray : public yasio::byte_buffer {
> public:
>   IndexArray(const std::initializer_list<unsigned int>& rhs) 
>   { 
>       this->assign((const uint8_t*)rhs.begin(), (const uint8_t*)rhs.end());
>   }
> 
>   IndexArray(const std::initializer_list<unsigned short>& rhs) 
>   { 
>       this->assign((const uint8_t*)rhs.begin(), (const uint8_t*)rhs.end());
>   }
> };
> ```

Acutually, the byte_buffer just a container to instead std::vector, because it can store any type of integer sequence, and we can ensure the memory model same with GPU, then we can copy it to GPU faster without a immediately container when index format is U_SHORT

Reply:
When you inherit from byte_buffer it says that it's a final class and that it cannot be inherited, what is the solution?

Reply:
will upgrade it later